File: leetcode/medium/MaximumDistanceToClosestPerson.py
from typing import List


class Solution:
    def maxDistToClosest(self, seats: List[int]) -> int:
        """
        Finds the maximum distance to the closest person in a row of seats
        in a single pass.
        """
        n = len(seats)
        max_dist = 0
        # `last_person` stores the index of the last person encountered.
        last_person = -1

        for i, seat in enumerate(seats):
            if seat == 1:
                if last_person == -1:
                    # Case 1: Empty seats at the beginning of the row.
                    # The max distance is the index of this first person.
                    max_dist = i
                else:
                    # Case 2: Empty seats between two people.
                    # The best position is in the middle of the gap.
                    distance = (i - last_person) // 2
                    max_dist = max(max_dist, distance)
                # Update the index of the last person seen.
                last_person = i

        # Case 3: Empty seats at the end of the row.
        # The distance is from the last person to the end of the array.
        trailing_dist = n - 1 - last_person
        max_dist = max(max_dist, trailing_dist)

        return max_dist

File: leetcode/medium/MaximumDistanceToClosestPerson.java
/**
 * Solution class for maximizing distance to the closest person.
 */
class Solution {

    /**
     * Calculates the maximum distance to the closest person in a row of seats.
     * * The problem handles three specific scenarios regarding empty seats:
     * 1. Left Edge: Empty seats from the start (index 0) to the first person.
     * 2. Middle: Empty seats between two people.
     * 3. Right Edge: Empty seats from the last person to the end of the array.
     *
     * @param seats An array representing seat occupancy (1 is occupied, 0 is empty).
     * @return The maximum closest distance to a person.
     */
    public int maxDistToClosest(int[] seats) {
        int lastPersonIndex = -1; // Tracks the index of the previously found person
        int maxDistance = 0;      // Tracks the global maximum distance found so far

        for (int i = 0; i < seats.length; i++) {
            // Check if the current seat is occupied
            if (seats[i] == 1) {
                if (lastPersonIndex == -1) {
                    // CASE 1: Leading Empty Seats
                    // If this is the first person we've seen, the distance is simply 
                    // the number of seats from the start (index 0) to here.
                    maxDistance = i;
                } else {
                    // CASE 2: Middle Empty Seats
                    // If we have seen a person before, the empty seats are between two people.
                    // The safest seat is exactly in the middle.
                    // Distance = (current_index - last_person_index) / 2
                    maxDistance = Math.max(maxDistance, (i - lastPersonIndex) / 2);
                }
                // Update the last seen person to the current index
                lastPersonIndex = i;
            }
        }

        // CASE 3: Trailing Empty Seats
        // After the loop, check the distance from the very last person found
        // to the physical end of the row.
        int trailingDistance = seats.length - 1 - lastPersonIndex;
        maxDistance = Math.max(maxDistance, trailingDistance);

        return maxDistance;
    }
}